dev.yaml

```diff
@@ -0,0 +1,21 @@
+services:
```

Comment:
We'd also need `SYS_PTRACE` capability to run debugger or ASAN builds inside the container. This can be added in `devcontainer.json` too.

Reply:
Where is the Dockerfile for https://quay.io/repository/389ds/devcontainer?  Should it be stored in an accessible place?   If I remove "dnf install libasan" the build fails with:

```
/usr/bin/ld: cannot find /usr/lib64/libasan.so.8.0.0: No such file or directory
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:5801: libsvrcore.la] Error 1
make[1]: Leaving directory '/389-ds-base'
make: *** [Makefile:3946: all] Error 2
```

Seems like it isn't really optional if [programmatically running the build](https://github.com/slominskir/389-ds-base/blob/e5b1dbb055b1e1e821830d64f2bece0181160c22/build.sh#L2).

Similarly, the "dscontainer -r" command fails without "nss-tools" with:

```
  File "/usr/lib64/python3.12/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib64/python3.12/subprocess.py", line 1953, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/usr/bin/certutil'
```

Similar for "openssl"

Reply:
Ah, sorry, it's here:  https://github.com/389ds/ds-container/tree/main/devcontainer.  Prob should add libasan, openssl, and nss-tools there?

Reply:
Also, IntelliJ won't launch the devcontainer unlesss ps is installed.  Then it will try to copy a backend app into the container.  Very buggy at the moment though.  So probably need:

`dnf install procps` in the Containerfile as well.

Reply:
Bind mounting the project files from the container into the local filesystem for the IDE to edit works reasonably well so I'm not sure I'd use the remote editing IDE feature at the moment.   Currently with IntelliJ it opens a new window and kinda works but complains that Java isn't installed inside the container:

![image](https://github.com/389ds/389-ds-base/assets/2642119/e71e0bf3-a3c6-4eeb-be44-3fcace0a2d0c)

Reply:
Also probably bad to do both at the same time.  Either need to remove bind mount else don't remote edit.   Not sure if that means devcontainer author's must choose up front as if volume is defined in compose it get's used.

Reply:
Testing running the .devcontainer in GitHub codespaces fails at the moment.  At first glance it appears the version of Compose inside codespaces is too old (< 2.24.0) because the dev.yaml compose file appears to choke it with log output:

```
validating /var/lib/docker/codespacemount/workspace/389-ds-base/dev.yaml: services.dirsrv.env_file.0 must be a string


2024-02-15 18:21:02.402Z: Exit code 15
2024-02-15 18:21:02.473Z: {"outcome":"error","message":"Command failed: docker-compose -f /var/lib/docker/codespacemount/workspace/389-ds-base/dev.yaml -f /var/lib/docker/codespacemount/.persistedshare/docker-compose.codespaces.yml --profile * config","description":"An error occurred retrieving the Docker Compose configuration."}
2024-02-15 18:21:02.477Z: Error: Command failed: docker-compose -f /var/lib/docker/codespacemount/workspace/389-ds-base/dev.yaml -f /var/lib/docker/codespacemount/.persistedshare/docker-compose.codespaces.yml --profile * config
2024-02-15 18:21:02.481Z:     at _g (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:462:889)
2024-02-15 18:21:02.486Z:     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
2024-02-15 18:21:02.489Z:     at async J$ (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:429:1484)
2024-02-15 18:21:02.491Z:     at async x$ (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:409:3165)
2024-02-15 18:21:02.493Z:     at async gAA (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:481:3833)
2024-02-15 18:21:02.496Z:     at async BC (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:481:4775)
2024-02-15 18:21:02.498Z:     at async xeA (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:614:11265)
2024-02-15 18:21:02.500Z:     at async UeA (/.codespaces/agent/bin/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:614:11006)
2024-02-15 18:21:02.502Z: devcontainer process exited with exit code 1
``` 

Could be something else, but running that compose file via Docker Compose works locally with latest Docker Desktop on Windows.

Reply:
Also worth mentioning codespaces is more of a gee wizz cool thing, but may not be of practical use anyways.  Every GitHub user has a very small amount of free codespace time, but as some have observed, once you run put and actually have to pay for it it may end up being [cheaper to buy a Macbook laptop](https://notes.alexkehayias.com/github-codespaces-is-too-slow-and-expensive-to-replace-local-development/) and run locally.  I'd rather develop locally anyways.   Again, why my initial PR ignored the spec.  Just being able to run a container locally via compose that is pre-loaded with the correct versions of tools to build and run an app is the real win.

Also worth mentioning our preloaded [devcontainer](https://notes.alexkehayias.com/github-codespaces-is-too-slow-and-expensive-to-replace-local-development/) probably should use tini or similar.  [Example](https://github.com/JeffersonLab/dirsrv/blob/ed4b8d21988cfc42004f4902ddea9c3768219585/Dockerfile#L20-L23).

Reply:
This week I will get my hands on Windows 11 laptop to test this. I believe primary use case is still local development, not cloud-first, although it's nice if it's supported.